x/bank/keeper/msg_server.go

```diff
@@ -104,3 +105,76 @@ func (k msgServer) MultiSend(goCtx context.Context, msg *types.MsgMultiSend) (*t
 
 	return &types.MsgMultiSendResponse{}, nil
 }
+
+func (k msgServer) Mint(goCtx context.Context, req *types.MsgMint) (*types.MsgMintResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	minter, err := sdk.AccAddressFromBech32(req.Minter)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, coin := range req.Amount {
+		denom := coin.Denom
+		mgr := k.getDenomManager(denom)
+		if mgr == nil {
+			return nil, fmt.Errorf("no denom manager for denom %s", denom)
+		}
+
+		err = mgr.OnMint(ctx, minter, coin)
+		if err != nil {
+			return nil, err
+		}
+
+		err = k.addCoins(ctx, minter, []sdk.Coin{coin})
+		if err != nil {
+			return nil, err
+		}
+
+		supply := k.GetSupply(ctx, denom)
+		supply = supply.Add(coin)
+		k.setSupply(ctx, supply)
+
+	}
+
+	ctx.EventManager().EmitEvent(
+		types.NewCoinMintEvent(minter, req.Amount),
```

Comment:
I think it's more clear to create many events for each denom minting (so moving the event to the loop, and do one event per iteration).